### PR TITLE
HPCC-30731 Explore effect of compression on transfer times

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
Typical output:
```
Algorithm(options)  || Comp(ms) || Deco(ms) || 200MB/s (w,r)   || 1GB/s (w,r)     || 5GB/s (w,r)     || Ratio [cLen]
                    ||          ||          || 2Gb/s           || 10Gb/s          || 50Gb/s          ||
       uncompressed ||        0 ||        0 ||  524( 524, 524) ||  104( 104, 104) ||   20(  20,  20) ||  1.00 [104857600]
AES(012345678901234 ||     1584 ||      580 || 2391(1811, 807) || 2209(1629, 625) || 2173(1593, 589) ||  2.31 [45490676]
             DIFF() ||       65 ||       28 ||  617( 589, 552) ||  197( 169, 132) ||  113(  85,  48) ||  1.00 [104857644]
              LZW() ||     1382 ||      400 || 2009(1609, 627) || 1827(1427, 445) || 1791(1391, 409) ||  2.31 [45490665]
            RDIFF() ||       63 ||       28 ||  615( 587, 552) ||  195( 167, 132) ||  111(  83,  48) ||  1.00 [104857644]
              FLZ() ||      280 ||      118 ||  655( 537, 375) ||  449( 331, 169) ||  408( 290, 128) ||  2.04 [51499131]
   LZ4HC(hclevel=3) ||      988 ||       46 || 1257(1211, 269) || 1078(1032,  90) || 1042( 996,  54) ||  2.35 [44651877]
   LZ4HC(hclevel=4) ||     1055 ||       47 || 1323(1276, 268) || 1146(1099,  91) || 1110(1063,  55) ||  2.37 [44332401]
   LZ4HC(hclevel=5) ||     1112 ||       46 || 1379(1333, 267) || 1202(1156,  90) || 1166(1120,  54) ||  2.37 [44208851]
   LZ4HC(hclevel=6) ||     1175 ||       46 || 1441(1395, 266) || 1265(1219,  90) || 1229(1183,  54) ||  2.37 [44171840]
   LZ4HC(hclevel=8) ||     1351 ||       47 || 1618(1571, 267) || 1442(1395,  91) || 1406(1359,  55) ||  2.38 [44127887]
  LZ4HC(hclevel=10) ||     5766 ||       45 || 6031(5986, 265) || 5855(5810,  89) || 5819(5774,  53) ||  2.38 [44088283]
   LZ4HC(hclevel=8) ||     1342 ||       48 || 1610(1562, 268) || 1434(1386,  92) || 1398(1350,  56) ||  2.38 [44127887]
            LZ4HC() ||     1540 ||       44 || 1804(1760, 264) || 1628(1584,  88) || 1592(1548,  52) ||  2.38 [44118521]
              LZ4() ||      163 ||       38 ||  473( 435, 310) ||  255( 217,  92) ||  211( 173,  48) ||  1.92 [54481575]
```
Added reviewers primarily to see the results.

File: testing/unittests/jlibtests.cpp [hunk 1]
```diff
@@ -3126,46 +3125,33 @@ class JlibCompressionTestsStress : public CppUnit::TestFixture
...
                 {
-                    compressor->write(ptr, rowSz);
-                    ptr += rowSz;
+                    testCompressor(handler, "hclevel=3", rowSz, src.length(), src.bytes(), false);
+                    testCompressor(handler, "hclevel=4", rowSz, src.length(), src.bytes(), false);
+                    testCompressor(handler, "hclevel=5", rowSz, src.length(), src.bytes(), false);
+                    testCompressor(handler, "hclevel=6", rowSz, src.length(), src.bytes(), false);
+                    testCompressor(handler, "hclevel=8", rowSz, src.length(), src.bytes(), false);
+                    testCompressor(handler, "hclevel=10", rowSz, src.length(), src.bytes(), false);
+                    testCompressor(handler, "hclevel=8", rowSz, src.length(), src.bytes(), true);
```
Comment (jakesmith) [inline testing/unittests/jlibtests.cpp:3152]: In the output table it looks like this is a duplicate test of LZ4HC(hclevel=8) ..
Results are very similar, perhaps not worth including.

**Review by jakesmith — APPROVED** [review]
@ghalliday - 1 minor comment, but looks good.

File: testing/unittests/jlibtests.cpp (hunk 1 above)
Comment (ghalliday) [reply testing/unittests/jlibtests.cpp]: Pushed and squashed that change.  I'll edit above to show new output.  Reducing the number of virtual calls helps a little, but avoiding the internal memcpy saves about 20% for LZ4.

File: testing/unittests/jlibtests.cpp [hunk 2]
```diff
@@ -3126,53 +3126,128 @@ class JlibCompressionTestsStress : public CppUnit::TestFixture
...
             {
-                compressed.clear();
                 ICompressHandler &handler = iter->query();
+                const char * type = handler.queryType();
                 //Ignore unusual compressors with no expanders...
-                if (strieq(handler.queryType(), "randrow"))
+                if (strieq(type, "randrow"))
                     continue;
-                Owned<ICompressor> compressor = handler.getCompressor(streq("AES", handler.queryType()) ? aesKey: nullptr);
+                const char * options = streq("AES", handler.queryType()) ? aesKey: "";
```
Comment (mckellyln) [inline testing/unittests/jlibtests.cpp:3144]: handler.queryType() or type ?

**Review by mckellyln — APPROVED** [review]
Looks fine, one trivial comment.

**Pull request merged**